File: src/utils/sanitize.ts
import { promises as fs } from 'fs';
import path from 'path';

export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ValidationError';
  }
}

// Allowed characters in lane names: alphanumeric, underscore, hyphen
const SAFE_LANE_PATTERN = /^[a-zA-Z][a-zA-Z0-9_-]*$/;

// Shell metacharacters to block
const DANGEROUS_CHARS = /[;&|`$(){}[\]<>!#*?~]/;

export function sanitizeLaneName(lane: string): string {
  if (!lane || typeof lane !== 'string') {
    throw new ValidationError('Lane name must be a non-empty string');
  }

  const trimmed = lane.trim();

  if (DANGEROUS_CHARS.test(trimmed)) {
    throw new ValidationError(`Lane name contains invalid characters: ${trimmed}`);
  }

  if (!SAFE_LANE_PATTERN.test(trimmed)) {
    throw new ValidationError(
      `Lane name must start with a letter and contain only letters, numbers, underscores, and hyphens: ${trimmed}`
    );
  }

  return trimmed;
}

export async function validatePath(inputPath: string): Promise<string> {
  if (!inputPath || typeof inputPath !== 'string') {
    throw new ValidationError('Path must be a non-empty string');
  }

  const trimmed = inputPath.trim();

  // Resolve to absolute path and normalize
  const resolved = path.resolve(trimmed);
  const normalized = path.normalize(resolved);

  // Check for path traversal (.. that escapes the intended directory)
  // Allow .. only if it stays within a reasonable path structure
  if (trimmed.includes('..')) {
    const originalParts = trimmed.split(path.sep).filter(p => p !== '');
    const normalizedParts = normalized.split(path.sep).filter(p => p !== '');

    // If normalized has fewer parts and original had .., potential traversal
    if (originalParts.includes('..') && normalizedParts.length < originalParts.filter(p => p !== '..').length) {
      throw new ValidationError(`Path contains suspicious traversal: ${trimmed}`);
    }
  }

  // Verify the path exists
  try {
    await fs.access(normalized);
  } catch {
    throw new ValidationError(`Path does not exist: ${normalized}`);
  }

  return normalized;
}

export async function validateProjectPath(projectPath: string): Promise<string> {
  const validated = await validatePath(projectPath);

  // Check it's a directory
  const stats = await fs.stat(validated);
  if (!stats.isDirectory()) {
    throw new ValidationError(`Project path must be a directory: ${validated}`);
  }

  return validated;
}

File: src/fastlane_mcp/utils/sanitize.py
"""Input sanitization and validation utilities."""

import re
from pathlib import Path

class ValidationError(Exception):
    """Raised when input validation fails."""
    pass


# Lane names must start with letter, contain only alphanumeric, underscore, hyphen
SAFE_LANE_PATTERN = re.compile(r'^[a-zA-Z][a-zA-Z0-9_-]*$')

# Characters that could be used for shell injection
DANGEROUS_CHARS = re.compile(r'[;&|`$(){}[\]<>!#*?~\\"\']')


def sanitize_lane_name(lane: str | None) -> str:
    """Validate and sanitize a fastlane lane name.

    Args:
        lane: The lane name to sanitize

    Returns:
        The sanitized lane name

    Raises:
        ValidationError: If the lane name is invalid
    """
    if not lane or not isinstance(lane, str):
        raise ValidationError("Lane name must be a non-empty string")

    trimmed = lane.strip()

    if not trimmed:
        raise ValidationError("Lane name must be a non-empty string")

    if DANGEROUS_CHARS.search(trimmed):
        raise ValidationError(f"Lane name contains invalid characters: {trimmed}")

    if not SAFE_LANE_PATTERN.match(trimmed):
        raise ValidationError(f"Invalid lane name format: {trimmed}")

    return trimmed


async def validate_project_path(path: str | None) -> Path:
    """Validate project path exists and is a directory.

    Args:
        path: The path to validate

    Returns:
        The resolved Path object

    Raises:
        ValidationError: If the path is invalid
    """
    if not path or not isinstance(path, str):
        raise ValidationError("Path must be a non-empty string")

    # Check for path traversal attempts
    if ".." in path:
        original_parts = path.split("/")
        if ".." in original_parts:
            raise ValidationError(f"Path contains suspicious traversal: {path}")

    p = Path(path).resolve()

    if not p.exists():
        raise ValidationError(f"Path does not exist: {path}")

    if not p.is_dir():
        raise ValidationError(f"Path is not a directory: {path}")

    return p
